crates/crypto/src/secp256k1.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{ecdh::EcdhKey, KeyError};
+use alloc::{vec, vec::Vec};
+use ring::hkdf;
+use sp_core::{ecdsa, Pair};
+
+pub type Signature = ecdsa::Signature;
+
+const SEED_SIZE: usize = 32;
+type Seed = [u8; SEED_SIZE];
+
+pub trait Signing {
+    fn sign_data(&self, data: &[u8]) -> Signature;
+
+    fn verify_data(&self, sig: &Signature, data: &[u8]) -> bool;
+}
+
+impl Signing for ecdsa::Pair {
+    fn sign_data(&self, data: &[u8]) -> Signature {
+        sp_core::Pair::sign(self, data)
+    }
+
+    fn verify_data(&self, sig: &Signature, data: &[u8]) -> bool {
+        ecdsa::Pair::verify(sig, data, &self.public())
+    }
+}
+
+// Phala Network Poc-4 Genesis Block Hash
+const KDF_SALT: [u8; 32] = [
+    0x18, 0xe8, 0x76, 0xad, 0xfa, 0x74, 0xcc, 0x74, 0x3b, 0x4b, 0x7d, 0x7f, 0x92, 0xc9, 0x6e, 0x03,
+    0xde, 0x55, 0x9a, 0x3c, 0x15, 0x27, 0x05, 0x81, 0xfc, 0xe4, 0x45, 0x96, 0x3d, 0x90, 0xf8, 0x5e,
+];
+
+pub trait KDF {
+    fn derive_secp256k1_pair(&self, info: &[&[u8]]) -> Result<ecdsa::Pair, KeyError>;
+
+    fn derive_ecdh_key(&self) -> Result<EcdhKey, KeyError>;
+}
+
+/// Generic newtype wrapper that lets us implement traits for externally-defined
+/// types.
+///
+/// Refer to https://github.com/briansmith/ring/blob/main/tests/hkdf_tests.rs
+#[derive(Debug, PartialEq)]
+struct My<T: core::fmt::Debug + PartialEq>(T);
+
+impl hkdf::KeyType for My<usize> {
+    fn len(&self) -> usize {
+        self.0
+    }
+}
+
+impl From<hkdf::Okm<'_, My<usize>>> for My<Vec<u8>> {
+    fn from(okm: hkdf::Okm<My<usize>>) -> Self {
+        let mut r = vec![0u8; okm.len().0];
+        okm.fill(&mut r).unwrap();
+        Self(r)
+    }
+}
+
+impl KDF for ecdsa::Pair {
+    fn derive_secp256k1_pair(&self, info: &[&[u8]]) -> Result<ecdsa::Pair, KeyError> {
```

Comment:
// TODO(shelven): allow to specify the salt from pruntime (instead of hard code)

Reply:
will be fixed